tools/xmlschema.py

```diff
@@ -0,0 +1,462 @@
+from xml.etree import ElementTree
+from copy import deepcopy
+from itertools import permutations, zip_longest
+
+from pathlib import Path
+import argparse
+from typing import List
+
+
+cmd_arg_parser = argparse.ArgumentParser(
+    description="Create an XML schema from a SDFormat file."
+)
+cmd_arg_parser.add_argument(
+    "-i",
+    "--in",
+    dest="source",
+    required=True,
+    type=str,
+    help="SDF file inside of directory to compile.",
+)
+cmd_arg_parser.add_argument(
+    "-s",
+    "--sdf",
+    dest="directory",
+    required=True,
+    type=str,
+    help="Directory containing all the SDF files.",
+)
+cmd_arg_parser.add_argument(
+    "-o",
+    "--out",
+    dest="target",
+    default=".",
+    type=str,
+    help="Output directory for xsd file. Will be created if it doesn't exit.",
+)
+cmd_arg_parser.add_argument(
+    "--ns-prefix",
+    dest="ns_prefix",
+    default="http://sdformat.org/schema",
+    type=str,
+    help="Prefix for generated xsd namespaces.",
+)
+
+args = cmd_arg_parser.parse_args()
+source_dir = Path(args.directory)
+source: Path = source_dir / args.source
+
+root = ElementTree.parse(source).getroot()
+declared = dict()
+declared["imports"] = set()
+
+
+def _to_simple_type(in_type: str) -> str:
+    known_types = {
+        "unsigned int": "xs:unsignedInt",
+        "unsigned long": "xs:unsignedLong",
+        "bool": "xs:boolean",
+        "string": "xs:string",
+        "double": "xs:double",
+        "int": "xs:int",
+        "float": "xs:float",
+        "char": "xs:char",
+        "vector3": "types:vector3",
+        "vector2d": "types:vector2d",
+        "vector2i": "types:vector2i",
+        "pose": "types:pose",
+        "time": "types:time",
+        "color": "types:color",
+    }
+
+    try:
+        return known_types[in_type]
+    except KeyError:
+        raise RuntimeError(f"Unknown simple type: {in_type}") from None
+
+
+# collect existing namespaces
```

Comment:
@Bi0T1N Hmm, I think this is mainly a stylistic choice. 

The `if __name__ == "__main__"` guard is useful if your have double-barreled files that - in parallel - function as both a script and as a module. Here, the purpose is to be a script only (there is no intention towards reusing the classes/functions elsewhere), so I chose not to use the guard. In my experience, that leads to cleaner scripts, because you can write (and debug) them sequentially rather than having to jump around the file, because classes live at the top and execution logic at the bottom.